Examples/Algorithms/TrackFinding/src/TrackFindingAlgorithm.cpp

```diff
@@ -152,51 +161,141 @@ ActsExamples::ProcessCode ActsExamples::TrackFindingAlgorithm::execute(
-          track, *pSurface, extrapolator, extrapolationOptions,
-          m_cfg.extrapolationStrategy, logger());
-      if (!extrapolationResult.ok()) {
-        m_nFailedExtrapolation++;
-        ACTS_ERROR("Extrapolation for seed "
-                   << iseed << " and track " << track.index()
-                   << " failed with error " << extrapolationResult.error());
-        continue;
-      }
+      std::size_t nSecond = 0;
 
       // Set the seed number, this number decrease by 1 since the seed number
-      seedNumber(track) = nSeed - 1;
+      // has already been updated
+      seedNumber(firstTrack) = nSeed - 1;
+
+      if (m_cfg.twoWay) {
+        std::optional<Acts::VectorMultiTrajectory::TrackStateProxy> firstState;
+        for (auto st : firstTrack.trackStatesReversed()) {
+          bool isMeasurement =
+              st.typeFlags().test(Acts::TrackStateFlag::MeasurementFlag);
+          bool isOutlier =
+              st.typeFlags().test(Acts::TrackStateFlag::OutlierFlag);
+          // We are excluding non measurement states and outlier here. Those can
+          // decrease resolution because only the smoothing corrected the very
+          // first prediction as filtering is not possible.
+          if (isMeasurement && !isOutlier) {
+            firstState = st;
+          }
+        }
+
+        if (firstState.has_value()) {
+          Acts::BoundTrackParameters secondInitialParameters(
+              firstState->referenceSurface().getSharedPtr(),
+              firstState->parameters(), firstState->covariance(),
+              firstInitialParameters.particleHypothesis());
+
+          auto secondResult = (*m_cfg.findTracks)(secondInitialParameters,
+                                                  secondOptions, tracksTemp);
+
+          if (!secondResult.ok()) {
+            ACTS_WARNING("Second track finding failed for seed "
+                         << iSeed << " with error" << secondResult.error());
+          } else {
+            auto firstFirstState =
+                std::next(firstTrack.trackStatesReversed().begin(),
+                          firstTrack.nTrackStates() - 1);
+
+            auto& secondTracksForSeed = secondResult.value();
+            for (auto& secondTrack : secondTracksForSeed) {
+              // TODO select track
+              if (secondTrack.nTrackStates() < 2) {
+                continue;
+              }
+
+              secondTrack.reverseTrackStates(true);
+              seedNumber(secondTrack) = nSeed - 1;
+
+              (*firstFirstState).previous() =
+                  (*std::next(secondTrack.trackStatesReversed().begin()))
+                      .index();
+              secondTrack.tipIndex() = firstTrack.tipIndex();
+
+              Acts::calculateTrackQuantities(secondTrack);
+
+              // TODO this extrapolation should not be necessary as the CKF will
+              // target this surface anyway and could communicate the parameters
```

Comment:
Why can't we remove this extrapolation now? Maybe explain in the `TODO` comment.

Reply:
That looked like a limitation of the CKF. Maybe this is gone now - I can try again. I think the problem was / is that an already complete track will not generate any new states in the second search and will not have the extrapolated parameters attached to a second track.

Reply:
either way, can you update or remove the comment?